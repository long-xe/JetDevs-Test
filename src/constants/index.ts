export {NAVIGATION, TABS} from './navigation';