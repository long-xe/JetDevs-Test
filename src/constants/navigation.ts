export const NAVIGATION = {
  login: 'Login Screen',
  home: 'Home Screen',
  favorite: 'Favorite Screen'
};

export const TABS = {
  home: 'Home',
  favorite: 'Favorite',
};
