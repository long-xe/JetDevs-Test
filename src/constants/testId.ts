
export const ScreenTestId = {
  LoginScreen:{
    emailInput:'email-input-login',
    passwordInput:'password-input-login'
  },
};

export const ComponentTestId = {
 
};
