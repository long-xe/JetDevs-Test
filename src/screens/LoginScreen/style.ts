import { StyleSheet } from "react-native";
import { windowDimensions } from "../../utils/dimension";
import { Colors, Radius, SizeFonts } from "../../styles";
import { scale } from "../../utils/scaling";

export const styles = StyleSheet.create({
  container: { flex: 1 },
  wrapper: {
    marginHorizontal: scale(20),
    flex: 1,
    backgroundColor: "white",
    justifyContent: "center",
  },
  imgWrapper: {
    height: 60,
    width: 60,
    position: "absolute",
    top: -30,
    left: windowDimensions.width / 2 - 50,
    borderRadius: 50,
    backgroundColor: "white",
    alignItems: "center",
    justifyContent: "center",
  },
  img: {
    resizeMode: "contain",
    width: 50,
    height: 50,
  },
  contentWrapper: {
    justifyContent: "space-around",
    height: "80%",
    alignItems: "center",
    paddingHorizontal: scale(30),
  },
  title: {
    fontSize: SizeFonts.h3,
    fontWeight: "600",
    textAlign: "center",
    color: Colors.black
  },
  inputWrapper: {
    width: "100%",
    gap: scale(30),
  },
  btn: {
    backgroundColor: Colors.primary,
    width: "100%",
    padding: scale(14),
    borderRadius: Radius.s,
  },
  btnText: {
    color: Colors.white,
    fontSize: SizeFonts.l,
    fontWeight: "700",
    textAlign: "center",
  },
  input: {
    borderWidth: 0,
    borderBottomWidth: 1,
  }
});
