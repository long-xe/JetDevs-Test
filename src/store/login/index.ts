import { reducer } from "./login.slice";

export default {
    reducer,
}