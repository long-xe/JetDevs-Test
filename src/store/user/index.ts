import { reducer } from "./user.slice";
import * as actions from './user.action'

export default {
    reducer,
    ...actions,
}