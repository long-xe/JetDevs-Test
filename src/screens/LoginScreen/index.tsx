import { View, Text, TouchableOpacity, Image } from 'react-native';
import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { LockLogo, MailLogo, MockLogo } from '../../../assets';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { Colors } from '../../styles';
import TextField from '../../components/inputs';
import { emailCheck, passwordCheck } from '../../utils/validation';
import type { RootState } from '../../store/store';
import { styles } from './style';
import { checkLoginState } from '../../store/login/login.slice';
import { GlobalStyle } from '../../styles/globalStyle';
import { ScreenTestId } from '../../constants/testId';

enum EFocusInput {
  None = 'none',
  Password = 'password',
  Email = 'email',
}

export default function LoginScreen() {
  const dispatch = useDispatch();
  const { isValidAccount } = useSelector((state: RootState) => state.login);
  const insets = useSafeAreaInsets();

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const [currentFocus, setCurrentFocused] = useState(EFocusInput.None);

  const [error, setError] = useState(false);

  const checkLogin = () => {
    if (!emailCheck(email).error && !passwordCheck(password).error) {
      setError(false);
      dispatch(checkLoginState({ email: email, password: password }));
    } else {
      setError(true);
    }
  };

  return (
    <View style={{ ...styles.container, marginVertical: insets.top + 60 }}>
      <View style={styles.wrapper}>
        <View style={styles.imgWrapper}>
          <Image source={MockLogo} resizeMode="contain" style={styles.img} />
        </View>
        <View style={styles.contentWrapper}>
          <Text style={styles.title}>LOGIN</Text>
          <View style={styles.inputWrapper}>
            <TextField
              testID={ScreenTestId.LoginScreen.emailInput}
              styleContainer={GlobalStyle.fullWidth}
              style={{
                ...styles.input,
                borderColor:
                  currentFocus === EFocusInput.Email
                    ? Colors.primary
                    : Colors.stroke,
              }}
              placeholder="Email"
              leftIcon={MailLogo}
              autoCapitalize="none"
              tintColor={
                currentFocus === EFocusInput.Email ? Colors.primary : undefined
              }
              onFocus={() => {
                setCurrentFocused(EFocusInput.Email);
              }}
              onBlur={() => {
                setCurrentFocused(EFocusInput.None);
              }}
              onChangeText={text => {
                setEmail(text);
              }}
              value={email}
              errorMessage={error ? emailCheck(email).errorMessage : undefined}
            />
            <TextField
              testID={ScreenTestId.LoginScreen.passwordInput}
              styleContainer={GlobalStyle.fullWidth}
              style={{
                ...styles.input,
                borderColor:
                  currentFocus === EFocusInput.Password
                    ? Colors.primary
                    : Colors.stroke,
              }}
              placeholder="Password"
              leftIcon={LockLogo}
              autoCapitalize="none"
              secureTextEntry={true}
              tintColor={
                currentFocus === EFocusInput.Password
                  ? Colors.primary
                  : undefined
              }
              onFocus={() => {
                setCurrentFocused(EFocusInput.Password);
              }}
              onBlur={() => {
                setCurrentFocused(EFocusInput.None);
              }}
              onChangeText={text => {
                setPassword(text);
              }}
              value={password}
              errorMessage={
                error
                  ? passwordCheck(password).errorMessage
                  : !isValidAccount
                  ? 'Login failed! Please check email or password!'
                  : undefined
              }
            />
          </View>

          <TouchableOpacity onPress={() => checkLogin()} style={styles.btn}>
            <Text style={styles.btnText}>LOGIN</Text>
          </TouchableOpacity>
        </View>
      </View>
    </View>
  );
}
