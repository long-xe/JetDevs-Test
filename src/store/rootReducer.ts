
import login from './login';
import user from './user';

export const rootReducer = {
  users: user.reducer,
  login: login.reducer
};
