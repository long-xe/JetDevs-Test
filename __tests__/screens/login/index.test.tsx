import { fireEvent, render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { act } from 'react-test-renderer';
import { ScreenTestId } from '../../../src/constants/testId';
import LoginScreen from '../../../src/screens/LoginScreen';
import { store } from '../../../src/store/store';

jest.mock('@react-navigation/native', () => ({
  useRoute: () => ({
    params: {},
  }),
  useNavigation: jest.fn(),
}));

jest.mock('@react-navigation/native-stack', () => ({
  createNativeStackNavigator: jest.fn(),
}));

jest.mock('@react-navigation/bottom-tabs', () => ({
  createBottomTabNavigator: jest.fn(),
}));

jest.mock('react-native-safe-area-context', () => {
  const inset = { top: 0, right: 0, bottom: 0, left: 0 }
  return {
    SafeAreaProvider: jest.fn().mockImplementation(({ children }) => children),
    SafeAreaConsumer: jest
      .fn()
      .mockImplementation(({ children }) => children(inset)),
    useSafeAreaInsets: jest.fn().mockImplementation(() => inset),
  }})

describe('main login screen', () => {
  it('should renders screen correctly', () => {
    const page = render(
        <Provider store={store}>
          <LoginScreen />
        </Provider>
    );
    expect(page).toMatchSnapshot();
  });

  it('input should type normally', () => {
    const { getByTestId } = render(
        <Provider store={store}>
          <LoginScreen />
        </Provider>
    );

    const emailInput = getByTestId(ScreenTestId.LoginScreen.emailInput);
    const passwordInput = getByTestId(ScreenTestId.LoginScreen.passwordInput);

    fireEvent.changeText(emailInput, 'test');
    fireEvent.changeText(passwordInput, 'test1');

    expect(emailInput.props.value).toBe('test');
    expect(passwordInput.props.value).toBe('test1');
  });
});
