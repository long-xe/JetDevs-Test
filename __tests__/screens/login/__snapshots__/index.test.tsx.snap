// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main login screen should renders screen correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "marginVertical": 60,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 50,
          "height": 60,
          "justifyContent": "center",
          "left": 325,
          "position": "absolute",
          "top": -30,
          "width": 60,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          {
            "testUri": "../../../assets/img/mock_logo.png",
          }
        }
        style={
          {
            "height": 50,
            "resizeMode": "contain",
            "width": 50,
          }
        }
      />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "height": "80%",
          "justifyContent": "space-around",
          "paddingHorizontal": 30,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 24,
            "fontWeight": "600",
            "textAlign": "center",
          }
        }
      >
        LOGIN
      </Text>
      <View
        style={
          {
            "gap": 30,
            "width": "100%",
          }
        }
      >
        <View
          style={
            [
              {
                "justifyContent": "center",
              },
              {
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#ffffff",
                  "borderColor": "#BFBFBF",
                  "borderRadius": 8,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 82.5,
                  "justifyContent": "center",
                  "paddingHorizontal": 15,
                },
                {
                  "borderBottomWidth": 1,
                  "borderColor": "#BFBFBF",
                  "borderWidth": 0,
                },
                {},
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 35.25,
                  "marginRight": 10,
                  "opacity": 1,
                  "resizeMode": "contain",
                  "width": 35.25,
                }
              }
            >
              <Image
                source={
                  {
                    "testUri": "../../../assets/img/mail.png",
                  }
                }
                style={
                  [
                    {
                      "height": "100%",
                      "width": "100%",
                    },
                    {
                      "tintColor": undefined,
                    },
                  ]
                }
              />
            </View>
            <TextInput
              autoCapitalize="none"
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Email"
              placeholderTextColor="#BFBFBF"
              style={
                {
                  "color": "#000000",
                  "flex": 1,
                  "fontSize": 15,
                  "height": "100%",
                }
              }
              testID="email-input-login"
              value=""
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "justifyContent": "center",
              },
              {
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#ffffff",
                  "borderColor": "#BFBFBF",
                  "borderRadius": 8,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 82.5,
                  "justifyContent": "center",
                  "paddingHorizontal": 15,
                },
                {
                  "borderBottomWidth": 1,
                  "borderColor": "#BFBFBF",
                  "borderWidth": 0,
                },
                {},
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 35.25,
                  "marginRight": 10,
                  "opacity": 1,
                  "resizeMode": "contain",
                  "width": 35.25,
                }
              }
            >
              <Image
                source={
                  {
                    "testUri": "../../../assets/img/lock.png",
                  }
                }
                style={
                  [
                    {
                      "height": "100%",
                      "width": "100%",
                    },
                    {
                      "tintColor": undefined,
                    },
                  ]
                }
              />
            </View>
            <TextInput
              autoCapitalize="none"
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              placeholderTextColor="#BFBFBF"
              secureTextEntry={true}
              style={
                {
                  "color": "#000000",
                  "flex": 1,
                  "fontSize": 15,
                  "height": "100%",
                }
              }
              testID="password-input-login"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#e82282",
            "borderRadius": 6,
            "opacity": 1,
            "padding": 14,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "#ffffff",
              "fontSize": 16,
              "fontWeight": "700",
              "textAlign": "center",
            }
          }
        >
          LOGIN
        </Text>
      </View>
    </View>
  </View>
</View>
`;
